### PR TITLE
add day 8

### PR DESCRIPTION
**Pull request by wtchangdm**
(definitely not idiomatic, just trying to get through this day)

### CODE REVIEW

File: 2023/08/wtchangdm/day8.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+struct Map {
+    cursor: usize,
+    directions: Vec<Direction>,
+    nodes: HashMap<String, [String; 2]>,
+}
+
+impl Map {
+    fn from(input: &[String]) -> Self {
+        let mut nodes: HashMap<String, [String; 2]> = HashMap::new();
```
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:27]: Unless you are intended to make it explicit, there is no need to write the type out here.

```suggestion
        let mut nodes = HashMap::new();
```

File: 2023/08/wtchangdm/day8.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        let mut iter = input.iter().peekable();
```
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:28]: There is no usage of `peek()`.

```suggestion
        let mut iter = input.iter();
```

File: 2023/08/wtchangdm/day8.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+struct Map {
+    cursor: usize,
+    directions: Vec<Direction>,
+    nodes: HashMap<String, [String; 2]>,
+}
+
+impl Map {
+    fn from(input: &[String]) -> Self {
```
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:26]: This is more like a constructor than a conversion. I may just call this `new` or some other thing instead.

```suggestion
    fn new(input: &[String]) -> Self {
```

File: 2023/08/wtchangdm/day8.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+pub fn solve_part2(input: &[String]) -> u64 {
+    let mut map = Map::from(input);
+    let mut starting_nodes: Vec<String> = map
+        .nodes()
+        .keys()
+        .filter(|node| node.ends_with('A'))
+        .cloned()
+        .collect();
+
+    let mut steps: Vec<u64> = vec![0; starting_nodes.len()];
```
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:97]: ditto

```suggestion
    let mut steps = vec![0; starting_nodes.len()];
```

File: 2023/08/wtchangdm/day8.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        Self {
+            directions,
+            nodes,
+            cursor: 0,
+        }
+    }
+
+    fn nodes(&self) -> &HashMap<String, [String; 2]> {
+        &self.nodes
+    }
```
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:57]: I feel like it's an unsuccessful abstraction. We have a `Map` type but still expose the entire `HashMap` implementation detail.

Not really a big deal for a puzzle solver :)

File: 2023/08/wtchangdm/day8.rs (hunk 3 above)
Comment (weihanglo) [inline 2023/08/wtchangdm/day8.rs:26]:
```suggestion
    fn from(input: &[&str]) -> Self {
```

We end up calling `to_string` or `into` to clone the String, so could we just make the type `&[&str]` so no double clones needed?

And the `input` in tests could be mininized to 

```rust
        let input = [
            "RL",
            "",
            "AAA = (BBB, CCC)",
            "BBB = (DDD, EEE)",
            "CCC = (ZZZ, GGG)",
            "DDD = (DDD, DDD)",
            "EEE = (EEE, EEE)",
            "GGG = (GGG, GGG)",
            "ZZZ = (ZZZ, ZZZ)",
        ];
```

File: 2023/08/wtchangdm/day8.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Copy)]
+enum Direction {
+    Left,
+    Right,
+}
+
+impl Direction {
+    fn from(ch: &char) -> Self {
```
[the remark this answers is not in the document]
Comment (wtchangdm) [reply 2023/08/wtchangdm/day8.rs:10]: Thanks. Not really familiar with some (probably pretty common) traits ATM.

File: 2023/08/wtchangdm/day8.rs (hunk 1 above)
Comment (wtchangdm) [reply 2023/08/wtchangdm/day8.rs:27]: Yeah, actually I use just trying to express the left and right direction but too lazy to bother with dedicated struct, hence the weird type here.

File: 2023/08/wtchangdm/day8.rs (hunk 2 above)
Comment (wtchangdm) [reply 2023/08/wtchangdm/day8.rs:28]: I was trying to use peek() at first and forgot to remove this as well. Thanks for spotting this.

File: 2023/08/wtchangdm/day8.rs (hunk 5 above)
Comment (wtchangdm) [reply 2023/08/wtchangdm/day8.rs:57]: True. Since these can just access the internals directly. Will revise along with other suggestions.

**Comment by wtchangdm** [comment]
@weihanglo Thanks again. I will find some time to refactor a little bit. Mark as draft for now.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
# 👍🏾

**Pull request merged**